src/bp/core/messaging/messaging-service.ts

```diff
@@ -0,0 +1,117 @@
+import { IO, MessagingConfig } from 'botpress/sdk'
+import { ConfigProvider } from 'core/config'
+import { EventEngine, Event } from 'core/events'
+import { TYPES } from 'core/types'
+import { inject, injectable, postConstruct } from 'inversify'
+import { AppLifecycle, AppLifecycleEvents } from 'lifecycle'
+import { MessagingClient } from './messaging-client'
+
+@injectable()
+export class MessagingService {
+  private clientSync!: MessagingClient
+  private clientsByBotId: { [botId: string]: MessagingClient } = {}
+  private botsByClientId: { [clientId: string]: string } = {}
+  private channelNames = ['messenger', 'slack', 'smooch', 'teams', 'telegram', 'twilio', 'vonage']
+
+  constructor(
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.ConfigProvider) private configProvider: ConfigProvider
+  ) {}
+
+  @postConstruct()
+  async init() {
+    this.eventEngine.register({
+      name: 'messaging.sendOut',
+      description: 'Sends outgoing messages to external messaging',
+      order: 20000,
+      direction: 'outgoing',
+      handler: this.handleOutgoingEvent.bind(this)
+    })
+
+    await AppLifecycle.waitFor(AppLifecycleEvents.STUDIO_READY)
+
+    this.clientSync = new MessagingClient(`http://localhost:${process.MESSAGING_PORT}`, process.INTERNAL_PASSWORD)
+  }
+
+  async loadMessagingForBot(botId: string) {
+    await AppLifecycle.waitFor(AppLifecycleEvents.STUDIO_READY)
+
+    const config = await this.configProvider.getBotConfig(botId)
+    let messaging = (config.messaging || {}) as MessagingConfig
+
+    const setupConfig = {
+      name: botId,
+      ...messaging
+      // We use the SPINNED_URL env var for now to force the messaging server to
+      // make its webhook requests to the process that started it.
+      // webhooks: [{ url: `${process.EXTERNAL_URL}/api/v1/chat/receive` }]
+    }
+
+    const { id, token } = await this.clientSync.syncClient(setupConfig)
+    let modified = false
+
+    if (id && id !== messaging.id) {
+      messaging = {
+        ...messaging,
+        id,
+        token
+      }
+      modified = true
+    }
+
+    if (modified) {
+      await this.configProvider.mergeBotConfig(botId, { messaging })
+    }
```

Comment:
I'm not sure to understand why the mergeBotConfig is not simply moved up in the condition instead of relying on the "modified" variable, am I missing something?

Reply:
There used to be multiple conditions that set the `modified` variable. I removed that now that it's not useful anymore